src/Chainweb/Payload/RestAPI.hs

```diff
@@ -142,6 +148,25 @@ payloadPostApi
     . Proxy (PayloadPostApi v c)
 payloadPostApi = Proxy
 
+-- | @POST \/chainweb\/\<ApiVersion\>\/\<InstanceId\>\/chain\/\<ChainId\>\/heightedpayload\/batch@
+--
+-- The query may return any number (including none) of the requested payload
+-- data. Results are returned in any order.
+--
+type HeightedPayloadPostApi_
+    = "payload"
+    :> "batch"
+    :> ReqBody '[JSON] [(BlockHeight, BlockPayloadHash)]
+    :> Post '[JSON] [PayloadData]
+
+type HeightedPayloadPostApi (v :: ChainwebVersionT) (c :: ChainIdT)
+    = 'ChainwebEndpoint v :> ChainEndpoint c :> HeightedPayloadPostApi_
+
+heightedPayloadPostApi
+    :: forall (v :: ChainwebVersionT) (c :: ChainIdT)
+    . Proxy (HeightedPayloadPostApi v c)
+heightedPayloadPostApi = Proxy
+
```

Comment:
What about just using the normal payload batch endpoint with optional block height property in the request body on a per payload basis? The implementation could do something like:

1. Collect and sort all request items that have a (non-null) value for block height
2. group collected items in continuous range.
3. Query each range group using an rocksdb iterator
4. Collect all request items that don't have a block height
5. (optional) order response items according to order of request items

For backward compatibility, the response would included in the block height (see comment above) if it was present in the request.

This approach would result in heterogenous arrays, which is fine in JSON. Alternatively, we could require that the block height is specified for either all or none of the items, which would avoid heterogenous arrays.

On the P2P API we should probably enforce the block height property on request items. On the service API it would be optional. Maybe even there we could deprecate the query by payload hash, because payload hashes are obtained from the block header which also includes the height.